packages/hash/api/src/model/knowledge/link.model.ts

```diff
@@ -111,19 +110,191 @@ export default class {
+      /** @todo: rely on Graph API to validate the index */
+      if (index < 0 || index > siblingLinks.length) {
+        throw new Error("Provided link index is out of bounds");
+      }
+
+      await Promise.all(
+        siblingLinks
+          .filter((sibling) => sibling.index! >= index)
+          .map((sibling) =>
+            sibling.updateWithoutUpdatingSiblings(graphApi, {
+              updatedIndex: sibling.index! + 1,
+              updatedBy: createdBy,
+            }),
+          ),
+      );
+    }
+
+    return await LinkModel.createLinkWithoutUpdatingSiblings(graphApi, {
+      ...params,
+      index,
+    });
+  }
+
+  private async updateWithoutUpdatingSiblings(
+    graphApi: GraphApi,
+    params: { updatedIndex: number; updatedBy: string },
+  ) {
+    const { updatedIndex, updatedBy } = params;
+
+    const { index: previousIndex } = this;
+
+    if (previousIndex === undefined) {
+      throw new Error("Cannot make an un-ordered link ordered");
+    }
+
+    if (previousIndex === updatedIndex) {
+      throw new Error("No-op: link already has this index");
+    }
+
+    /**
+     * @todo: call dedicated Graph API method to update the index of a link instead of re-creating the link manually
+     * @see https://app.asana.com/0/1202805690238892/1203031430417465/f
+     */
+    await this.removeWithoutUpdatingSiblings(graphApi, {
+      removedBy: updatedBy,
+    });
+
+    const updatedLink = await LinkModel.createLinkWithoutUpdatingSiblings(
+      graphApi,
+      {
+        ...this,
+        index: updatedIndex,
+        createdBy: updatedBy,
+      },
+    );
+
+    return updatedLink;
+  }
+
+  /**
+   * Update the link
+   *
+   * @param params.updatedIndex - the updated index of the link
+   * @param params.updatedBy - the account updating the link
+   */
+  async update(
+    graphApi: GraphApi,
+    params: { updatedIndex: number; updatedBy: string },
+  ) {
+    const { updatedIndex, updatedBy } = params;
+
+    const { index: previousIndex, linkTypeModel } = this;
+
+    const updatedLink = await this.updateWithoutUpdatingSiblings(graphApi, {
+      updatedIndex,
+      updatedBy,
+    });
+
+    const siblingLinks = await this.sourceEntityModel.getOutgoingLinks(
+      graphApi,
+      { linkTypeModel },
+    );
+
+    if (previousIndex === undefined) {
+      throw new Error("Cannot make an un-ordered link ordered");
+    }
+
+    // Whether the index of the link is being increased
+    const isIncreasingIndex = updatedIndex > previousIndex;
+
+    // The minimum index of the affected sibling links
+    const affectedSiblingsMinimumIndex = isIncreasingIndex
+      ? previousIndex + 1
+      : updatedIndex;
+
+    // The maximum index of the affected sibling links
+    const affectedSiblingsMaximumIndex = isIncreasingIndex
+      ? updatedIndex
+      : previousIndex - 1;
+
+    const affectedSiblings = siblingLinks.filter(
+      (sibling) =>
+        sibling.index! >= affectedSiblingsMinimumIndex &&
```

Comment:
Both of these being equalities (when usually one end of a range isn't inclusive) is confusing my brain a little but I _think_ it is correct. @thehabbos007 can you have a think about this as well and double-check

Reply:
> I wonder if it would be good to somehow unit test this in isolation outside integration tests so we can get more confidence about the indexing (although I appreciate the integration tests likely cover testing this)

If you feel strongly about it happy to write more tests around this. The main reason I haven't done so is because I've treated this as a temporary implementation, the workspace backend should not be where this is implemented.

Reply:
I don't feel strongly about it for this, but for any implementation on the rust side that mimicks this should definitely be tested well.